pkg/controller/direct/kms/keyhandle/keyhandle_controller.go

```diff
@@ -0,0 +1,199 @@
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting KeyHandle", "name", a.id.External)
+
+	req := &kmspb.GetKeyHandleRequest{Name: a.id.External}
+	keyhandlepb, err := a.gcpClient.GetKeyHandle(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting KeyHandle %q: %w", a.id.External, err)
+	}
+
+	a.actual = keyhandlepb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating KeyHandle", "name", a.id.External)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := KMSKeyHandleSpec_ToProto(mapCtx, &desired.Spec, a.id)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	parent, err := a.id.Parent()
+	if err != nil {
+		return err
+	}
+	req := &kmspb.CreateKeyHandleRequest{
+		Parent:    parent.String(),
+		KeyHandle: resource,
+	}
+	op, err := a.gcpClient.CreateKeyHandle(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating KeyHandle %s: %w", a.id.External, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("KeyHandle %s waiting creation: %w", a.id.External, err)
+	}
+	log.V(2).Info("successfully created KeyHandle", "name", a.id.External)
+
+	status := &krm.KMSKeyHandleStatus{}
+	status.ObservedState = KMSKeyHandleStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = &a.id.External
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update operation not supported for KeyHandle.
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	return fmt.Errorf("update operation not supported for KeyHandle resource")
+}
+
+func (a *Adapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.KMSKeyHandle{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(KMSKeyHandleSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	parent, err := a.id.Parent()
+	if err != nil {
+		return nil, err
+	}
+	obj.Spec.ProjectRef = &refs.ProjectRef{Name: parent.ProjectID}
+	obj.Spec.Location = direct.LazyPtr(parent.Location)
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+// Delete operation not supported for KeyHandle, so this operation is a no-op.
+func (a *Adapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("delete operation on KeyHandle resource not supported", "name", a.id.External)
```

Comment:
If user sending the same CREATE requests, will the KMS server create a new object everytime?

Reply:
as discussed, yes.